### PR TITLE
Refactor to use the params pattern

### PR DESCRIPTION
**Pull request by petems**
* Inherit from params in the base class ()
* Changes params to fail if directly referenced
* Requires major version bump: breaking behaviour
* Closes https://github.com/voxpupuli/puppet-mrepo/issues/42

### CODE REVIEW

**Comment by bastelfreak** [comment]
thanks for the PR @petems! Can you take a look at the merge conflicts?

File: manifests/init.pp [hunk 1]
```diff
@@ -28,29 +115,91 @@
...
+) inherits ::mrepo::params {
+
+  validate_re($source, '^git$|^package$')
+  validate_re($git_proto, '^git$|^https$')
+  validate_re($priority, '^\d+$')
+  validate_re($port, '^\d+$')
+  validate_bool($rhn)
+  validate_hash($descriptions)
+
+  assert_private('You can no longer refer to the mrepo::params class directly, see https://github.com/voxpupuli/puppet-mrepo#caveats for more details')
```
Comment (alexjfisher) [inline manifests/init.pp]: `assert_private` in the base class??

File: README.markdown [hunk 2]
```diff
@@ -162,6 +162,39 @@ page](https://github.com/dagwieers/mrepo/blob/master/docs/usage.txt).
 
 ## Caveats ##
 
+### Params Pattern ###
+
+This module was previously using a pattern
```
Comment (alexjfisher) [inline README.markdown]: This sentence isn't finished?

File: manifests/init.pp [hunk 3]
```diff
@@ -28,29 +115,91 @@
...
+  }
+  if $rhnget_download_all != undef {
+    validate_bool($rhnget_download_all)
+  }
+
+  # Validate selinux usage. If manually set, validate as a bool and use that value.
+  # If undefined and selinux is present and not disabled, use selinux.
+  case $mrepo::selinux {
+    undef: {
+      case $::selinux {
```
Comment (petems) [inline manifests/init.pp]: This is old enough that the selinux core fact probably didn't exist....

File: manifests/init.pp (hunk 1 above)
Comment (petems) [inline manifests/init.pp]: Fixed 👍

File: README.md [hunk 4]
```diff
@@ -162,6 +162,39 @@ page](https://github.com/dagwieers/mrepo/blob/master/docs/usage.txt).
 
 ## Caveats ##
 
+### Params Pattern ###
+
+This module was previously using a pattern
```
Comment (rnelson0) [inline README.md]: Seems to be missing a few words here. Not sure this paragraph is needed though.

File: manifests/init.pp [hunk 5]
```diff
@@ -28,29 +115,90 @@
 #
 # Copyright 2011 Puppet Labs, unless otherwise noted
```
Comment (rnelson0) [inline manifests/init.pp]: I know it's not directly related but should probably update this as well.

File: manifests/package.pp [hunk 6]
```diff
@@ -18,13 +18,12 @@
 #
 class mrepo::package {
 
-  include ::mrepo::params
```
Comment (rnelson0) [inline manifests/package.pp]: Should this inherit, or is it a private class now?

**Comment by rnelson0** [comment]
@petems found one or two issues, but mostly just had questions related to how to convert. Thanks!

File: manifests/package.pp (hunk 6 above)
Comment (petems) [inline manifests/package.pp]: Private 👍

File: README.md (hunk 4 above)
Comment (petems) [inline README.md]: Yeah, it's a big change so I thought I'd mention it 👍

**Comment by alexjfisher** [comment]
@petems Good work getting this done and merged.  I guess we should do a final puppet 3 release and then anyone who's interested can start making puppet 4 only changes?

In particular, I noticed [this](https://docs.puppet.com/puppet/latest/style_guide.html#parameter-defaults) today! ;)

**Comment by bastelfreak** [comment]
@alexjfisher the final puppet3 release got released today.

**Pull request merged**